### ISSUE TITLE
Switching between stereo/mono render modes, toggling undistortion in NDK

### ISSUE THREAD

**Issue by scorpeeon** [googlevr/gvr-android-sdk#353]
In NDK, in order to be able to render in normal mono 2D mode (and distortion correction disabled) I'm looking for functionality similar to this in the Java API:
https://developers.google.com/vr/android/reference/com/google/vr/sdk/base/GvrView#setStereoModeEnabled(boolean)

But I found no reference to something like this in the NDK SDK documentation.
If it's really missing, what should be the easiest way to achieve this?

Of course I can set the Viewport to cover the whole screen and only call the render method for a single eye - but even then if I draw to the frame (acquired by swapchain_->AcquireFrame()) the image will still have undistortion applied and will be split into 2 parts.
Is there any way to disable undistortion - or is the only way to use a different framebuffer, not the one the SDK provides?

**Comment by jdduke** [comment]
When you're running in mono 2D mode, do you actually need any behavior provided by GvrLayout or the NDK? If you don't (i.e., you don't need distortion rendering or head tracking), my first suggestion would be to swap out GvrLayout for whatever kind of View you need to properly render in 2D.

Note that there's an "unofficial" API to disable some stereo mode features, via GvrLayout.setStereoModeEnabled(). However, all this does is disable the UI overlay and prevent any VR entry screens from showing; you'd still need to customize your render path to avoid the distortion path. Without async reprojection, this should be fairly straightforward, as you can simply bind the default framebuffer and render as usual. However, *with* async reprojection, that approach will not work, and there's actually no good workaround that allows you to preserve GvrLayout's position in your View hierarchy. This is actually a known issue with GvrView, that it doesn't actually support toggling distortion correction when async reprojection is enabled. We'll see about addressing this in a future release, and will post here with more details.

**Comment by scorpeeon** [comment]
Great, thanks for the quick reply!

So in my specific case, the app would basically need to have 2 modes: normal VR mode (with head tracking, undistortion and async reprojection all enabled) and normal 2D mode (without all these things). There's no need to switch between them runtime, so I would most likely do that by 2 specific product flavors. (also the issue you mentioned about toggling undistortion shouldn't be a big problem in this case)

As far as I know, GvrLayout is needed to be able to create a native GVR context, so swapping that out would need some fundamental changes to the code basically removing most (or all) GVR specific functionality from that build flavor. It might be less problematic as I imagine it now, but I'd go first with a more simple approach if it's possible (as the simple method I mentioned in the Java API).

Yeah, I think the simplest way would be just using the default framebuffer, I'll try and see how it goes. Thanks!

**Comment by sigmaxipi** [comment]
For the 2D mode, do you need any Daydream functionality such as head-tracking or controller input? If not, the easiest option is to use a separate Android Activity that has no GVR code but calls your app's renderer in monoscopic mode. If you only need head tracking, then you could use Android's sensor APIs rather than GVR's APIs.

The code for your 2D Activity would be something like this:

Activity.java:
```
    // Configure sensors and touch in the Activity constructor.
    sensorManager = (SensorManager) context.getSystemService(Context.SENSOR_SERVICE);
    // TYPE_ROTATION_VECTOR is the easiest sensor since it handles all the hard math for fusion.
    orientationSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR);
    phoneOrientationListener = new PhoneOrientationListener();
  }

  @Override
  public void onResume() {
    super.onResume();
    // Use the fastest sensor readings.
    sensorManager.registerListener(
        phoneOrientationListener, orientationSensor, SensorManager.SENSOR_DELAY_GAME);
  }
```

Handle sensor events:
```
  // Detect sensor events and save them as a matrix.
  private class PhoneOrientationListener implements SensorEventListener {
    @Override
    public void onSensorChanged(SensorEvent event) {
      synchronized (/* you need to synchronize between this event thread and the GL thread */) {
        // Android coordinate system assumes Y points North and Z points to the sky. OpenGL has
        // Y pointing up and Z pointing toward the user.
        SensorManager.getRotationMatrixFromVector(phoneInWorldSpaceMatrix, event.values);
        // Rotate from Android coordinates to OpenGL coordinates.
        Matrix.rotateM(phoneInWorldSpaceMatrix, 0, 90, 1, 0, 0);
      }
    }
  }
```

Transform GL camera based on sensor data. If you're using the NDK for rendering, pass this matrix to your native code.
```
public void onDrawFrame(GL10 unused) {
  ...
  synchronized (...) {
    Matrix.multiplyMM(viewMatrix, 0, phoneInWorldSpaceMatrix, 0, cameraMatrix, 0);
  }
  ...
}
```

**Comment by jdduke** [comment]
@scorpeeon I'm going to close this out per the latest comments, feel free to file a new issue with additional needs, thanks.

**Comment by scorpeeon** [comment]
@jduke yeah no problem. your comment was very helpful, I could make it work as you mentioned just by binding the default framebuffer. One other issue I saw was that first the depth buffer didn't work properly, then I realized that was because in the GVR sample apps the GLSurfaceView is initialized with a depth size of 0 (surfaceView.setEGLConfigChooser(8, 8, 8, 0, 0, 0)) but after using a GLSurfaceView with the correct parameters and 16 bit depth size, it all worked as expected.

Thanks for the help again.
